parsing/builtin_attributes.ml

```diff
@@ -15,6 +15,72 @@
 
 open Asttypes
 open Parsetree
+open Ast_helper
+
+
+module Attribute_table = Hashtbl.Make (struct
+  type t = string with_loc
+
+  let hash : t -> int = Hashtbl.hash
+  let equal : t -> t -> bool = (=)
+end)
+let unused_attrs = Attribute_table.create 128
+let mark_used t = Attribute_table.remove unused_attrs t
+
+(* [attr_order] is used to issue unused attribute warnings in the order the
+   attributes occur in the file rather than the random order of the hash table
+*)
+let attr_order a1 a2 =
+  match String.compare a1.loc.loc_start.pos_fname a2.loc.loc_start.pos_fname
+  with
+  | 0 -> Int.compare a1.loc.loc_start.pos_cnum a2.loc.loc_start.pos_cnum
+  | n -> n
+
+let warn_unused () =
+  let keys = List.of_seq (Attribute_table.to_seq_keys unused_attrs) in
+  let keys = List.sort attr_order keys in
+  List.iter (fun sloc ->
+    Location.prerr_warning sloc.loc (Warnings.Misplaced_attribute sloc.txt))
+    keys
+
+(* These are the attributes that are tracked in the builtin_attrs table for
+   misplaced attribute warnings. *)
+let builtin_attrs =
+  [ "alert"; "ocaml.alert"
+  ; "boxed"; "ocaml.boxed"
+  ; "deprecated"; "ocaml.deprecated"
+  ; "deprecated_mutable"; "ocaml.deprecated_mutable"
+  ; "explicit_arity"; "ocaml.explicit_arity"
+  ; "immediate"; "ocaml.immediate"
+  ; "immediate64"; "ocaml.immediate64"
+  ; "inline"; "ocaml.inline"
+  ; "inlined"; "ocaml.inlined"
+  ; "noalloc"; "ocaml.noalloc"
+  ; "ppwarning"; "ocaml.ppwarning"
+  ; "tailcall"; "ocaml.tailcall"
+  ; "unboxed"; "ocaml.unboxed"
+  ; "untagged"; "ocaml.untagged"
+  ; "unrolled"; "ocaml.unrolled"
+  ; "warnerror"; "ocaml.warnerror"
+  ; "warning"; "ocaml.warning"
+  ; "warn_on_literal_pattern"; "ocaml.warn_on_literal_pattern"
+  ]
+
+let builtin_attrs =
+  let tbl = Hashtbl.create 128 in
+  List.iter (fun attr -> Hashtbl.add tbl attr ()) builtin_attrs;
+  tbl
+
+let is_builtin_attr s = Hashtbl.mem builtin_attrs s
+
+type attr_tracking_time = Parser | Invariant_check
+
+let register_attr attr_tracking_time name =
+  match attr_tracking_time with
+  | Parser when !Clflags.all_ppx <> [] -> ()
+  | Parser | Invariant_check ->
+    if is_builtin_attr name.txt then
+      Attribute_table.replace unused_attrs name ()
```

Comment:
I think there are a couple ways to read this remark, depending on whether you meant (1) "(duplicate attribute) warning" or (2) "duplicate (attribute warning)" .  Let me answer both :)

1. The logic for `Warnings.Duplicated_attribute` (warning 54) is meant to be unchanged by this PR.  In particular, it doesn't use this table at all - it looks for multiple copies of the attribute on the term, which will still be there.  So I think there is no change.

2. The other question would be how many times we issue warning 53 if a ppx adds multiple copies of an attribute with the same location in an incorrect spot.  I believe the old answer is 0 (even in cases where w53 otherwise worked well) and the new answer is 1.  I say the old answer is 0 because the old implementation looked for attributes using [find_attribute](https://github.com/ocaml/ocaml/blob/f6d6f65f0fe985101e7455a60a025a693a5f212e/lambda/translattribute.ml#L49), which would return no attributes at all in the case of a duplicate (this is the small bug I mentioned fixing in commit 3).

As for point 2, we could imagine issuing warning 53 multiple times for the same location by recording in the hashtbl how many times this attribute has been seen at this location.  Or adjusting the one warning to mention multiple copies were found, with the same technique.  This does not seem worth doing, to me, for the corner case of a broken ppx - but I'm happy to add it if it seems useful to you.

Reply:
I was thinking of case 2, but I agree that having one warning emitted is probably the right behavior.